src/extensions/scratch3_vizblocks/index.js

```diff
@@ -136,6 +155,197 @@ class Scratch3VizBlocks {
+            this.runtime.requestRedraw();
+        }
+
+        if (!this._posEmpty){
+            this._xPos.length = 0;
+            this._yPos.length = 0;
+            this._posEmpty = true;
+        }
+
+    }
+
+    /**
+     * The pen "pen down" block causes the target to leave pen trails on future motion.
+     * @param {object} args - the block arguments.
+     * @param {object} util - utility object provided by the runtime.
+     */
+    drawAxis (args, util) {
+        const target = util.target;
+        
+        // console.log('draw axis');
+        target.setVisible(false);
+
+        const penSkinId = this._getPenLayerID();
+        if (penSkinId >= 0) {
+            const penState = this._getPenState(target);
+            target.x = this._xCenter + this._width;
+            target.y = this._yCenter;
+            this.runtime.renderer.penLine(penSkinId, penState.penAttributes, this._xCenter, this._yCenter, target.x, target.y);
+
+            for (let i = 0; i * this._interval < this._width; i++){
+                const interval = i * this._interval;
+                this.runtime.renderer.penLine(penSkinId, penState.penAttributes, this._xCenter + interval, this._yCenter, this._xCenter + interval, this._yCenter + 5);
+                this.drawText(this._xCenter + interval, this._yCenter - 10, i, target, false);
+            }
+            
+            // console.log(target.getCostumes());
+            target.x = this._xCenter;
+            target.y = this._yCenter + this._height;
+            this.runtime.renderer.penLine(penSkinId, penState.penAttributes, this._xCenter, this._yCenter, target.x, target.y);
+            for (let i = 1; i * this._interval < this._height; i++){
+                const interval = i * this._interval;
+                this.runtime.renderer.penLine(penSkinId, penState.penAttributes, this._xCenter, this._yCenter + interval, this._xCenter + 5, this._yCenter + interval);
+                this.drawText(this._xCenter - 10, this._yCenter + interval, i, target, true);
+            }
+            
+            // console.log(target.getCostumes().length);
+            for (let i = 0, j = 1; i * this._interval < this._width && j * this._interval < this._height; i++, j++){
+                this.drawText(this._xCenter + (i * this._interval), this._yCenter - 10, i, target, true);
+                this.drawText(this._xCenter - 10, this._yCenter + (j * this._interval), j, target, true);
+            }
+            // console.log(target.getCostumes().length);
+            
+            // target.setVisible(false);
+            this.runtime.requestRedraw();
+
+        }
+    }
+
+    drawText (xPos, yPos, text, target, flag){
+        const index = text;
+        const runtime = this.runtime;
+
+        // console.log('draw text');
+        const costumeObject = new Object();
+        if (!flag){
+            runtime.storage.load(AssetType.ImageVector, assetID[index]).then(asset => {
```

Comment:
I think this block of code should be abstracted into another function because:
1. It isn't drawing any text as the function name suggests.
2. It is loading assets based on the flag that is supplied. Since this is an asynchronous behavior, the assets might not have finished loading before this function finishes executing. How about we load the costumes at the start of the program?

Reply:
1. This block has been abstracted into function loadCostumes()
2. Once block 'draw axis X' or 'draw axis Y ' is clicked, this function will be run first